private/optimalScaling.m

```diff
@@ -0,0 +1,291 @@
+function [varargout] = optimalScaling(varargin)
+% optimalScaling() computes the optimal scaling parameters.
+%
+% USAGE:
+% * [...]  = optimalScaling(iy,simulation,D,options)
+% * [...]  = optimalScaling(iy,simulation,D,options,scale)
+% * [s,ds] = optimalScaling(...)
+%
+% Parameters:
+%  iy: observable index for which the scaling parameter and the
+%      noise should be computed
+%  for the other inputs see loglikelihoodHierarchical.m
+%
+% Return values:
+%   s: the optimal scaling parameter for observable i
+%   ds: (1 x 1 x n_theta x n_r) matrix with the derivatives of the scaling parameters
+
+iy = varargin{1};
+simulation = varargin{2};
+D = varargin{3};
+options = varargin{4};
+
+if nargin == 5
+    scale = varargin{5};
+else
+    scale = 'lin';
+end
+
+%% CHECK SCENARIO
+for iiy = 1:numel(iy)
+    if(strcmp(options.noise{iy(iiy)},'multiple') && ...
+            strcmp(options.scaling{iy(iiy)},'single'))
+        error(['In options.(',num2str(iy(iiy)),...
+            ') you combined noise:multiple and scaling:single which is not valid.']);
+    end
+end
+% all data points that are used for the calculation of a scaling
+% parameters need to have the same noise parameter!
+
+scaling = options.scaling{iy(1)};
+noise = options.noise{iy(1)};
+for iiy = 1:numel(iy)
+    if ~strcmp(options.noise{iy(iiy)},noise) || ...
+            ~strcmp(options.scaling{iy(iiy)},scaling)
+        if ~strcmp(options.scaling{iy(iiy)},'absolute')
+            error('different assumptions')
+        end
+    end
+end
+
+%% INITIALIZATION OF DIMENSIONS
+n_e = size(D,2); %number of experiments
+n_r = size(D(1).my,3); %number of replicates
+if nargout > 1
+    n_theta = size(simulation(1).sy,3); %number of parameters
+end
+%% COMPUTATION OF s AND ds
+s = zeros(1,1,n_r); %scaling parameters
+if nargout > 1
+    ds = zeros(1,1,n_theta,n_r); %derivatives of the scaling parameters
+end
+
+try
+    switch scaling
+        case 'absolute'
+            s = ones(1,1,n_r);
+            if nargout > 1
+                ds = zeros(1,1,n_theta,n_r);
+            end
+        case 'multiple'
+            switch options.distribution
```

Comment:
For the normal distribution the only difference between `multiple` and `single` i could spot was whether `s = exp(bsxfun(@rdivide,logmy-logy,multfact));` or `s = bsxfun(@times,ones(1,1,n_r),exp((logmy-logy)/multfact));` was called in the end. I believe here is a lot of room to optimize code. This should make it easier to extend the code to other distributions in the future.